File: src/main/java/com/pedigree/editor/commands/AddIndividualCommand.java
package com.pedigree.editor.commands;

import com.pedigree.editor.Command;
import com.pedigree.model.Individual;
import com.pedigree.storage.ProjectRepository;

import java.util.Objects;

public class AddIndividualCommand implements Command {
    private final ProjectRepository.ProjectData data;
    private final Individual individualToAdd;
    private boolean added;

    public AddIndividualCommand(ProjectRepository.ProjectData data, Individual individualToAdd) {
        this.data = Objects.requireNonNull(data, "data");
        this.individualToAdd = Objects.requireNonNull(individualToAdd, "individualToAdd");
    }

    @Override
    public void execute() {
        boolean exists = data.individuals.stream().anyMatch(i -> i.getId().equals(individualToAdd.getId()));
        if (exists) {
            throw new IllegalStateException("Individual with id already exists: " + individualToAdd.getId());
        }
        data.individuals.add(individualToAdd);
        added = true;
    }

    @Override
    public void undo() {
        if (!added) { return; }
        data.individuals.removeIf(i -> i.getId().equals(individualToAdd.getId()));
        added = false;
    }
}

File: src/main/kotlin/com/family/tree/editor/commands/AddIndividualCommand.kt
package com.family.tree.editor.commands

import com.family.tree.editor.Command
import com.family.tree.model.Individual
import com.family.tree.storage.ProjectRepository
import java.util.Objects

class AddIndividualCommand(
    private val data: ProjectRepository.ProjectData,
    private val individualToAdd: Individual
) : Command {
    private var added: Boolean = false

    init {
        Objects.requireNonNull(data, "data")
        Objects.requireNonNull(individualToAdd, "individualToAdd")
    }

    override fun execute() {
        val exists = data.individuals.any { it.id == individualToAdd.id }
        if (exists) {
            throw IllegalStateException("Individual with id already exists: ${individualToAdd.id}")
        }
        data.individuals.add(individualToAdd)
        added = true
    }

    override fun undo() {
        if (!added) return
        data.individuals.removeIf { it.id == individualToAdd.id }
        added = false
    }
}
